source/common/http/conn_manager_impl.cc

```diff
@@ -1632,7 +1639,10 @@ void ConnectionManagerImpl::ActiveStream::recreateStream(
             filter_state->parent(), StreamInfo::FilterState::LifeSpan::FilterChain);
   }
 
-  new_stream.decodeHeaders(std::move(request_headers_), true);
+  new_stream.decodeHeaders(std::move(request_headers_), !proxy_body);
+  if (proxy_body) {
+    new_stream.decodeData(*request_data, true);
```

Comment:
So when we only supported requests-without-body it was header only requests and you could infer the request was complete.

Did you mean to only support redirects with a full body?   I'm not sure if the remaining "safe to redirect" checks check that.  I could imagine an upstream being able to determine the request should be redirected from headers to a large post, where Envoy had not received the entire body yet so if you're willing to support it, let's do that, and if not let's comment that it's not supported, and either way integration test :-)

Reply:
Maybe I'm not understanding correctly, but wouldn't that be prevented by the `downstream_end_stream_` check:
https://github.com/envoyproxy/envoy/blob/fcdaf20519216082db374dfad92e5b4771cc37fd/source/common/router/router.cc#L1488

I think if there are still bytes being read from downstream then that would be false and we wouldn't attempt a redirect?

But yes either way I'll add an integration test!

Reply:
yeah, what I was saying is I don't think this code works with partial body.  I forgot it's already reflected as point 3) in the redirect docs, and in the router :-)

Let's then comment for other potentially concerned people just above decodeData that we know end_stream is true because redirects are only supported when the full request is read 

Bonus points if you do an integration test where the upstream tries to send a retry when a partial body has been sent, so we can verify the code "fails" as expected.